### PR TITLE
Detect `unused-asyncio-dangling-task` (`RUF006`) on unused assignments

### LINKED ISSUE #8863: Detect `asyncio-dangling-task` (`RUF006`) when discarding return value
```python
import asyncio

for i in range(10):
    asyncio.create_task(foo)
```

triggers `asyncio-dangling-task` but

```python
import asyncio

for i in range(10):
    _ = asyncio.create_task(foo)
```

does not.

I would expect Ruff to trigger `asyncio-dangling-task` when the return value of `asyncio.create_task` is discarded, following the [rationale of the rule](https://docs.astral.sh/ruff/rules/asyncio-dangling-task/).

### PR DESCRIPTION
**Pull request by asafamr-mm**
## Summary

Fixes #8863 : Detect asyncio-dangling-task (RUF006) when discarding return value 

## Test Plan

added new two testcases, changed result of an old one that was made more specific

### CODE REVIEW

File: crates/ruff_linter/src/checkers/ast/analyze/statement.rs [hunk 1]
```diff
@@ -1560,7 +1560,9 @@ pub(crate) fn statement(stmt: &Stmt, checker: &mut Checker) {
                 pylint::rules::named_expr_without_context(checker, value);
             }
             if checker.enabled(Rule::AsyncioDanglingTask) {
-                ruff::rules::asyncio_dangling_task(checker, value);
+                let mut diagnostics: Vec<Diagnostic> = vec![];
+                ruff::rules::asyncio_dangling_task(checker, value, &mut diagnostics);
+                checker.diagnostics.extend(diagnostics);
```
Comment (zanieb) [inline crates/ruff_linter/src/checkers/ast/analyze/statement.rs:1565]: Why this change? This doesn't match the pattern we use to add a diagnostic for other rules.

Comment (charliermarsh) [reply crates/ruff_linter/src/checkers/ast/analyze/statement.rs:1565]: Probably because it _is_ the pattern we use for the deferred scopes diagnostics, since they can't take a `mut` on `Checker` :)

I'd probably prefer changing the rule to return `Option<Diagnostic>` though, and then pushing onto the appropriate vector. That should work fine for both call-sites.

File: crates/ruff_linter/src/rules/ruff/rules/asyncio_dangling_task.rs [hunk 2]
```diff
@@ -105,6 +110,27 @@ pub(crate) fn asyncio_dangling_task(checker: &mut Checker, expr: &Expr) {
...
+        .filter(|binding| binding.kind.is_assignment() && !binding.is_used())
+    {
+        let Some(source) = binding.source else {
+            continue;
+        };
+        let Stmt::Assign(ast::StmtAssign { value, .. }) = checker.semantic().statement(source)
+        else {
+            continue;
+        };
+        asyncio_dangling_task(checker, value, diagnostics);
```
Comment (charliermarsh) [inline crates/ruff_linter/src/rules/ruff/rules/asyncio_dangling_task.rs:130]: Nice job figuring this out.

Comment (charliermarsh) [reply crates/ruff_linter/src/rules/ruff/rules/asyncio_dangling_task.rs:130]: Can you try using `typing::analyze::resolve_assignment` here instead? It will also handle `AnnAssign`.

**Review by charliermarsh — COMMENTED** [review]
Nice job! A few small comments. Thank you so much for taking this on.

File: crates/ruff_linter/src/checkers/ast/analyze/statement.rs (hunk 1 above)
Comment (zanieb) [reply crates/ruff_linter/src/checkers/ast/analyze/statement.rs:1565]: :D Don't listen to me listen to Charlie.

Comment (asafamr-mm) [reply crates/ruff_linter/src/checkers/ast/analyze/statement.rs]: 👍 changed to `Option<Diagnostic>`

File: crates/ruff_linter/src/rules/ruff/rules/asyncio_dangling_task.rs (hunk 2 above)
Comment (asafamr-mm) [reply crates/ruff_linter/src/rules/ruff/rules/asyncio_dangling_task.rs]: Hmm, not sure how to do it without some refactoring.
I currently use assignment value expression which (if I'm not mistaken) is discarded after processing in `resolve_assignment`?

For now I've extended the pattern match to include AnnAssign as done in [unnecessary_list_cast](https://github.com/asafamr-mm/ruff/blob/00bcfc87c8b5d1380f4902d4019205b2840caac6/crates/ruff_linter/src/rules/perflint/rules/unnecessary_list_cast.rs#L120 ) - but would gladly change it

**Review by charliermarsh — APPROVED** [review]
Thanks again!

**Comment by hauntsaninja** [comment]
I think this PR has caused ruff to start complaining even when assignments are to globals and nonlocals. Is that intentional?

**Comment by charliermarsh** [comment]
@hauntsaninja - probably not — if you share an example I’m happy to ensure it’s fixed in the next release.

**Comment by hauntsaninja** [comment]
Thanks, filed https://github.com/astral-sh/ruff/issues/9262 !
